### ISSUE TITLE
Scopes not working on HasManyThrough relationship

### ISSUE THREAD

**Issue by jcsoriano** [laravel/framework#33555]
- Laravel Version: 7.20.0
- PHP Version: 7.4.5
- Database Driver & Version: mysql 5.7.29

### Description:
Let's say I have a hasManyThrough relationship as follows:

```
User -> Post -> Comment
```

Post has a scope called `published()`

In other types of relationships like `hasMany`, I can use scopes:

`$post->comments()->published()->get()`

But if the relationship is `hasManyThrough`, it throws an error:

`$user->comments()->published()->get()`

```
exception: "BadMethodCallException"
file: "/project/vendor/laravel/framework/src/Illuminate/Support/Traits/ForwardsCalls.php"
line: 50
message: "Call to undefined method Illuminate\Database\Eloquent\Relations\HasManyThrough::published()"
```

### Steps To Reproduce:

Edit: See second comment for correct configuration (https://github.com/laravel/framework/issues/33555#issuecomment-660455991)

Let me know if you need any other information

**Comment by driesvints** [comment]
Please share the migrations.

**Comment by jcsoriano** [comment]
Hi @driesvints , there are two more information that may help debug this:

1. It's a polymorphic hasManyThrough
2. We are using UUIDs

Do you think any of the above would affect it?

You may disregard my original issue and see the information in this new comment instead as it is exact to what we have, I just changed the table and model names to not expose the actual schema of my company. The "middle" model (Post) is polymorphic, it can belong to either a Team or a User. 

**Team model and migration**
```
class Team
{
    public function posts()
    {
        return $this->morphMany(Post::class);
    }
}

Schema::create('teams', function (Blueprint $table) {
    $table->uuid('id');
    $table->string('name');
    $table->primary('id');
});
```

**User model and migration**
```
class User
{
    public function posts()
    {
        return $this->morphMany(Post::class);
    }
}

Schema::create('users', function (Blueprint $table) {
    $table->uuid('id');
    $table->string('name');
    ...
    $table->primary('id');
});
```

**Post model and migration**
```
class Post
{
    public function ownerable()
    {
        return $this->morphTo();
    }
}

Schema::create('posts', function (Blueprint $table) {
    $table->uuid('id');
    $table->string('ownerable_type');
    $table->uuid('ownerable_id');
    $table->string('name');
    $table->longText('content');
    $table->primary('id');
});
```

**Comment model and migration**
```
class Comment
{
    public function post()
    {
        return $this->belongsTo(Post::class);
    }
}

Schema::create('comments', function (Blueprint $table) {
    $table->uuid('id');
    $table->uuid('post_id');
    $table->longText('content');
    $table->primary('id');
});
```

**Comment by staudenmeir** [comment]
Please share the correct `Team::posts()` relationship.

**Comment by myckhel** [comment]
In your User model
`return $this->hasMany(Post::class, User::class)`
Is the second argument of `hasMany` not meant to be related column name?

**Comment by driesvints** [comment]
Closing this issue because it's inactive, already solved, old or not relevant anymore. Feel free to reply if you're still experiencing this issue and we'll re-open this issue.

**Comment by jcsoriano** [comment]
Hi @staudenmeir and @myckhel yes please disregard my first comment and refer to my second comment instead for the proper configuration. It still doesn't work

**Comment by taylorotwell** [comment]
I don't think it works because we haven't supported it yet. It's ambiguous. Is the scope supposed to apply to the far model or the in between model? You're trying to apply it to the in between model but I would expect it to apply to the child model.

**Comment by jcsoriano** [comment]
Hi @taylorotwell , sorry for the ambiguity in my report. I am applying the `published` scope to the child model instead of the in-between model. I have updated my second comment to reflect this more clearly

**Comment by jcsoriano** [comment]
Hi @taylorotwell and @driesvints this still isn't fixed, hope we can reopen the issue. Let me know what other information is needed to debug this. Thank you!

**Comment by dkulyk** [comment]
```php
    public function scopePublished(Builder $query)
    {
        return $query->whereNotNull('published_at');
    }
```

**Comment by jcsoriano** [comment]
Hi @dkulyk as far as I know you do not need to return the query object in scopes. I have been using scopes without returning the query object and it works on all instances

**Comment by staudenmeir** [comment]
@dkulyk was pointing out that your scope method needs to be named `scopePublished()`, not `published()`.

**Comment by dkulyk** [comment]
https://laravel.com/docs/7.x/eloquent#local-scopes

**Comment by jcsoriano** [comment]
Hi @staudenmeir and @dkulyk , so sorry I had to rewrite the code for this issue since I couldn't share the actual code of my company. In my actual code it is correctly `scopePublished`, I updated my second comment to reflect this. Unfortunately, it is still not working

**Comment by dkulyk** [comment]
Where is your comments relation for User?

**Comment by jcsoriano** [comment]
Hi @dkulyk I've updated the second comment to include the relation, and I'm pasting it as well here for your convenience:

```
    public function comments()
    {
        return $this->hasManyThrough(Comment::class, Post::class, 'ownerable_id')
            ->where('ownerable_type', 'users');
    }
```

Unfortunately I cannot do `->toSql()` because of the error of this issue:
<img width="803" alt="image" src="https://user-images.githubusercontent.com/5937317/88516137-6547f680-d01f-11ea-95dd-bcf4e241111a.png">

However, `$post->comments()->published()->toSql();` works:
```
select * from `comments` where `comments`.`post_id` = ? and `comments `.`post_id ` is not null and `published_at` is not null order by `created_at` desc
```

**Comment by dkulyk** [comment]
Maybe you are using any custom traits or anything else?

**Comment by jcsoriano** [comment]
I only have two traits in the middle polymorphic model:

- `SetsUuidWhenCreating` which sets a UUID to the id field when `::creating`
- and [Spatie's `HasTags`](https://github.com/spatie/laravel-tags) 

On the end model (Comment) it only has `SetsUuidWhenCreating`

User model there are a few but I tried removing all of the traits and the error is the same 🤔

**Comment by staudenmeir** [comment]
Can you reproduce the issue on a fresh Laravel installation?

**Comment by jcsoriano** [comment]
Hi @dkulyk and @staudenmeir , thank you very much for helping me debug this. It turned out to be a veeeery stupid mistake. I just flipped the order of the first 2 parameters in my `->hasManyThrough()` call. Alas, the joys of programming.

If only I could just copy the actual code you might have spotted it too, but since I had to recreate the logic when reporting it here (to keep my company's code confidential) I subconsciously fixed it while writing the issue here.

Thank you again for the time you spent helping out on this issue.
